--- 9.98/feq/tsfcom.cmn
+++ 10.61/feq/tsfcom.cmn
@@ -1,10 +1,10 @@
-C**********************************************************************
+C***********************************************************************
       COMMON/TSFCOM/TSFBUF, BJTIME, TSFRAT, BWYR, BORD, TRIBA, LENTRY, 
      A              NDFVAL, TSFDSN, NBUFF, NLUSE, SEFLAG, DTEND, LHEAD, 
-     B              GSTART, GEND, TSFTYP, LAGTSF, TAFAC
+     B              GSTART, GEND, TSFTYP, LAGTSF, TAFAC, dtsf_rec, frcst
 
       INTEGER LENTRY, NDFVAL, TSFDSN, NBUFF, NLUSE, SEFLAG, LHEAD,
-     A        TSFTYP, LAGTSF
+     A        TSFTYP, LAGTSF, dtsf_rec, frcst
       INTEGER BWYR(MNDBUF), BORD(MNDBUF),  
      A        GEND(0:MXGAGE), GSTART(0:MXGAGE)
       REAL DTEND
