--- 9.98/fequtl/conduit.for
+++ 10.61/fequtl/conduit.for
@@ -46,7 +46,7 @@
  
 C     + + + EXTERNAL NAMES + + +
       INTEGER LENSTR
-      EXTERNAL INL80, KIL, GET_INTERNAL_TAB_NUMBER, LENSTR,
+      EXTERNAL inline, KIL, GET_INTERNAL_TAB_NUMBER, LENSTR,
      A         STRIP_L_BLANKS
  
 C     + + + OUTPUT FORMATS + + +
@@ -66,7 +66,7 @@
       ELSE
         MXSLOT = 0.02134
       ENDIF
-      CALL INL80
+      CALL inline
      I          (STDIN, STDOUT,
      O           LINE)
       READ(LINE,'(7X,A)',ERR=991) TABID
@@ -87,7 +87,7 @@
  
 C     INPUT THE ADJUSTMENT FACTOR FOR THE BUILT IN EXTRAPOLATION
  
-      CALL INL80
+      CALL inline
      I          (STDIN, STDOUT,
      O           LINE)
       READ(LINE,'(A7,F10.0)',ERR=991) NAME, FACTOR
@@ -233,6 +233,7 @@
       INCLUDE 'arsize.prm'
       INCLUDE 'xscomu.cmn'
       INCLUDE 'nrdzcm.cmn'
+      INCLUDE 'xtadd.cmn'
  
 C     + + + SAVED VALUES + + +
       INTEGER NURQ
@@ -255,7 +256,12 @@
      B     SPAN(PMXSUB), WSLOT, XOFF, XT(23), YOFF, YT(23), ZEPS,
      C     ZMUD(PMXSUB), ZMUD2, ZOFF
       CHARACTER BETOPT*8, CIN*64, CZEPS*10, LINE*80, MONTON*8, OUTOPT*8,
-     A          SAVOPT*8, TYPE(PMXSUB)*10, TABID*16
+     A          SAVOPT*8, TYPE(PMXSUB)*10,
+     a khflag(pmxpnt)*1, alphaflag(pmxpnt)*1, betaflag(pmxpnt)*1,
+     b     maflag(pmxpnt)*1, mqflag(pmxpnt)*1,
+     c     zone*8, hgrid*8, vdatum*8, unitsys*8, basis*8
+
+
  
 C     + + + INTRINSICS + + +
       INTRINSIC ABS, MAX, MIN
@@ -265,7 +271,7 @@
       CHARACTER GETTOK*10
  
 C     + + + EXTERNAL NAMES + + +
-      EXTERNAL CXSTAB, GETTOK, INL80, INL80B, MKBOX, MKPIPE, RHARCH, 
+      EXTERNAL CXSTAB, GETTOK, inline, inlineb, MKBOX, MKPIPE, RHARCH, 
      A         RHMAK,  SETMUD, SETOPT, SETSUB, TABCHK, TABOUT, URQMAK,
      B         URQTE, LENSTR, READ_TABID_PLUS
  
@@ -429,7 +435,7 @@
       ELSE
         CFAC = 12.0
       ENDIF 
-      CALL INL80
+      CALL inline
      I          (STDIN, STDOUT,
      O           LINE)
       CALL READ_TABID_PLUS
@@ -458,7 +464,16 @@
  
       TABU = TAB
  
-      CALL INL80
+c     Get location items that may be present. If they are not present
+c     they will be set to default values.  The default requests FEQUTL
+c     to omit the items.  
+      call GET_lctn_ITEMS(STDIN, STDOUT,
+     M                   EFLAG)
+      call  SET_lctn_ITEMS(
+     O             zone, hgrid, vdatum, unitsys, basis, easting,
+     O             northing)
+
+      CALL inline
      I          (STDIN, STDOUT,
      O           LINE)
       READ(LINE,6,ERR=991) WSLOT
@@ -466,13 +481,13 @@
  
       SLOT = WSLOT
  
-      CALL INL80
+      CALL inline
      I          (STDIN, STDOUT,
      O           LINE)
       READ(LINE,6,ERR=991) HSLOT
       WRITE(STDOUT,58) HSLOT
  
-      CALL INL80
+      CALL inline
      I          (STDIN, STDOUT,
      O           LINE)
       READ(LINE,7,ERR=991) NPIPES, CZEPS
@@ -493,6 +508,9 @@
         EFLAG = 1
         RETURN
       ENDIF
+
+C     26 October 2006, ddf: Set the gisid to blank.  Uninitialized otherwise!
+      gisid = ' '
  
 C     Compute the controlling parameter for input of the conduit
 C     descriptions.
@@ -503,31 +521,31 @@
         JE = JS + 5
         JE = MIN(JE, NPIPES)
  
-        CALL INL80
+        CALL inline
      I            (STDIN, STDOUT,
      O             LINE)
         READ(LINE,4,ERR=991) (TYPE(J), J=JS,JE)
         WRITE(STDOUT,54) (TYPE(J), J=JS,JE)
  
-        CALL INL80
+        CALL inline
      I            (STDIN, STDOUT,
      O             LINE)
         READ(LINE,8,ERR=991) (SPAN(J),J=JS,JE)
         WRITE(STDOUT,70) (SPAN(J), J=JS,JE)
  
-        CALL INL80
+        CALL inline
      I            (STDIN, STDOUT,
      O             LINE)
         READ(LINE,8,ERR=991) (RISE(J),J=JS,JE)
         WRITE(STDOUT,71) (RISE(J), J=JS,JE)
  
-        CALL INL80
+        CALL inline
      I            (STDIN, STDOUT,
      O             LINE)
         READ(LINE,9,ERR=991) (BOTTOM(J), J=JS,JE)
         WRITE(STDOUT,72) (BOTTOM(J), J=JS,JE)
  
-        CALL INL80
+        CALL inline
      I            (STDIN, STDOUT,
      O             LINE)
         READ(LINE,10,ERR=991) (ROUGH(J), J=JS,JE)
@@ -536,7 +554,7 @@
 C       Look ahead for the mudline input line.  If not there, backspace
 C       the input file and set the mud line level to null.
  
-        CALL INL80B
+        CALL inlineb
      I            (STDIN, STDOUT,
      O             LINE)
         IF(LINE(1:4).NE.'MUDL') THEN
@@ -549,7 +567,7 @@
 C         READ THE MUDLINE LEVEL
           READ(LINE,12,ERR=991) (ZMUD(J), J=JS,JE)
           WRITE(STDOUT,96) (ZMUD(J), J=JS,JE)
-          CALL INL80
+          CALL inline
      I              (STDIN, STDOUT,
      O               LINE)
           READ(LINE,14,ERR=991) (NMUD(J), J=JS,JE)
@@ -623,7 +641,11 @@
             WRITE(STDOUT,86) SPAN(I), RISE(I)
  
             MAXSOF = MAX(MAXSOF, RISE(I) + ZOFF)
- 
+            if(i == 1) then
+               xoff = xoff + 0.5*span(i) + 0.25
+            else
+               xoff = xoff + 0.5*(span(i-1) + span(i)) + 0.25
+            endif
             CALL MKPIPE
      I                 (XOFF, ZOFF, NSIDES, SPAN(I), WSLOT, HSLOT,
      O                  NPNT, XU(ISTART), ZU(ISTART))
@@ -667,6 +689,11 @@
  100        CONTINUE
  
             YOFF = ZOFF + 0.4*D
+            if(i == 1) then
+               xoff = xoff + 0.5*span(i) + 0.25
+            else
+               xoff = xoff + 0.5*(span(i-1) + span(i)) + 0.25
+            endif
             CALL URQMAK
      I                 (WSLOT, HSLOT, YOFF, XOFF, NURQ, XT, YT,
      O                  XU(ISTART), ZU(ISTART), NPNT)
@@ -709,6 +736,11 @@
  102        CONTINUE
  
             YOFF = ZOFF + 0.63*D
+            if(i == 1) then
+               xoff = xoff + 0.5*span(i) + 0.25
+            else
+               xoff = xoff + 0.5*(span(i-1) + span(i)) + 0.25
+            endif
             CALL URQMAK
      I                 (WSLOT, HSLOT, YOFF, XOFF, NURQ, XT, YT,
      O                  XU(ISTART), ZU(ISTART), NPNT)
@@ -735,6 +767,11 @@
  
             IF(EFLAG.GT.0) RETURN
             YOFF = YOFF + ZOFF
+            if(i == 1) then
+               xoff = xoff + 0.5*span(i) + 0.25
+            else
+               xoff = xoff + 0.5*(span(i-1) + span(i)) + 0.25
+            endif
             CALL RHMAK
      I                (WSLOT, HSLOT, YOFF, XOFF, NRH, XT, YT,
      O                 XU(ISTART), ZU(ISTART), NPNT)
@@ -760,6 +797,11 @@
  
             IF(EFLAG.GT.0) RETURN
             YOFF = YOFF + ZOFF
+            if(i == 1) then
+               xoff = xoff + 0.5*span(i) + 0.25
+            else
+               xoff = xoff + 0.5*(span(i-1) + span(i)) + 0.25
+            endif
             CALL RHMAK
      I                (WSLOT, HSLOT, YOFF, XOFF, NRH, XT, YT,
      O                 XU(ISTART), ZU(ISTART), NPNT)
@@ -786,6 +828,11 @@
  
             IF(EFLAG.GT.0) RETURN
             YOFF = YOFF + ZOFF
+            if(i == 1) then
+               xoff = xoff + 0.5*span(i) + 0.25
+            else
+               xoff = xoff + 0.5*(span(i-1) + span(i)) + 0.25
+            endif
             CALL RHMAK
      I                (WSLOT, HSLOT, YOFF, XOFF, NRH, XT, YT,
      O                 XU(ISTART), ZU(ISTART), NPNT)
@@ -811,6 +858,11 @@
  
             IF(EFLAG.GT.0) RETURN
             YOFF = YOFF + ZOFF
+            if(i == 1) then
+               xoff = xoff + 0.5*span(i) + 0.25
+            else
+               xoff = xoff + 0.5*(span(i-1) + span(i)) + 0.25
+            endif
             CALL RHMAK
      I                (WSLOT, HSLOT, YOFF, XOFF, NRH, XT, YT,
      O                 XU(ISTART), ZU(ISTART), NPNT)
@@ -838,6 +890,11 @@
  
             IF(EFLAG.GT.0) RETURN
             YOFF = YOFF + ZOFF
+            if(i == 1) then
+               xoff = xoff + 0.5*span(i) + 0.25
+            else
+               xoff = xoff + 0.5*(span(i-1) + span(i)) + 0.25
+            endif
             CALL RHMAK
      I                (WSLOT, HSLOT, YOFF, XOFF, NRH, XT, YT,
      O                 XU(ISTART), ZU(ISTART), NPNT)
@@ -864,6 +921,11 @@
  
             IF(EFLAG.GT.0) RETURN
             YOFF = YOFF + ZOFF
+            if(i == 1) then
+               xoff = xoff + 0.5*span(i) + 0.25
+            else
+               xoff = xoff + 0.5*(span(i-1) + span(i)) + 0.25
+            endif
             CALL RHMAK
      I                (WSLOT, HSLOT, YOFF, XOFF, NRH, XT, YT,
      O                 XU(ISTART), ZU(ISTART), NPNT)
@@ -890,6 +952,11 @@
  
             IF(EFLAG.GT.0) RETURN
             YOFF = YOFF + ZOFF
+            if(i == 1) then
+               xoff = xoff + 0.5*span(i) + 0.25
+            else
+               xoff = xoff + 0.5*(span(i-1) + span(i)) + 0.25
+            endif
             CALL RHMAK
      I                (WSLOT, HSLOT, YOFF, XOFF, NRH, XT, YT,
      O                 XU(ISTART), ZU(ISTART), NPNT)
@@ -910,6 +977,11 @@
      O                 NURQ, XT, YT)
  
             YOFF = ZOFF + RISE(I)/2.0
+            if(i == 1) then
+               xoff = xoff + 0.5*span(i) + 0.25
+            else
+               xoff = xoff + 0.5*(span(i-1) + span(i)) + 0.25
+            endif
             CALL URQMAK
      I                 (WSLOT, HSLOT, YOFF, XOFF, NURQ, XT, YT,
      O                  XU(ISTART), ZU(ISTART), NPNT)
@@ -926,6 +998,11 @@
             WRITE(STDOUT,86) SPAN(I), RISE(I)
             MAXSOF = MAX(MAXSOF, RISE(I) + ZOFF)
  
+            if(i == 1) then
+               xoff = xoff + 0.5*span(i) + 0.25
+            else
+               xoff = xoff + 0.5*(span(i-1) + span(i)) + 0.25
+            endif
             CALL MKBOX
      I                (XOFF, ZOFF, SPAN(I), RISE(I), WSLOT, HSLOT,
      O                 NPNT, XU(ISTART), ZU(ISTART))
@@ -942,7 +1019,14 @@
  
           CALL SETSUB
      I               (I, NPNT, ISTART, ROUGH(I),
-     O                SBU, LSNU(I))
+     O                SBU, LSNU)
+c          write(stdout,91) i, type(i), npnt, istart, rough(i)
+c91    format(' i=',i5,' type=',a8,' npnt=',i5,' istart=',i5,
+c     a        ' rough(i)=',f10.4)
+c          do j=istart,istart+npnt-1
+c            write(stdout,89) j, xu(j), zu(j), lsnu(j)
+c89    format(i5,f10.4,f10.4,f10.3)
+c          enddo
  
           IF(ZMUD(I).GT.0.0) THEN
 C           ADJUST THE CROSS SECTION FOR A MUD LINE
@@ -956,9 +1040,11 @@
  
           ISTART = ISTART + NPNT
           NPNTU = NPNTU + NPNT
-          IF(I.NE.NPIPES) THEN
-            XOFF = XOFF + MAX(SPAN(I), SPAN(I+1)) + 1.0
-          ENDIF
+c          IF(I.NE.NPIPES) THEN
+c            write(stdout,8232) i, span(i), span(i+1)
+c8232  format('i=',i5,' span(i)=',f10.4,' span(i+1)=',f10.4)
+c            XOFF = XOFF + MAX(SPAN(I), SPAN(I+1)) + 1.0
+c          ENDIF
  500    CONTINUE
  
       IF(MAXSOF.GT.HSLOT) THEN
@@ -967,14 +1053,19 @@
         HSLOT = MAXSOF + 5.0
         GOTO 9000
       ENDIF
- 
-C      WRITE(STDOUT,*) ' '
-C      WRITE(STDOUT,*) ' DUMP OF CROSS-SECTION BOUNDARY'
-C      WRITE(STDOUT,*) ' '
-C      DO 1000 I=1,NPNTU
-C        WRITE(STDOUT,90) XU(I), ZU(I), SBU(I), LSNU(I)
-C90    FORMAT(F10.4,F10.3,I5,F10.3)
-C1000  CONTINUE
+
+c     Set sinuosity to 1.0 everywhere
+      do i=1,npntu
+        snu(i) = 1.0
+      end do
+ 
+c      WRITE(STDOUT,*) ' '
+c      WRITE(STDOUT,*) ' DUMP OF CROSS-SECTION BOUNDARY'
+c      WRITE(STDOUT,*) ' NPNTU=',npntu
+c      DO 1000 I=1,NPNTU
+c        WRITE(STDOUT,90) XU(I), ZU(I), SBU(I), LSNU(I)
+c90    FORMAT(F10.4,F10.3,I5,F10.3)
+c1000  CONTINUE
  
       NAVMU = 0
       NSUBU = NPIPES
@@ -1042,9 +1133,18 @@
           XSTU(J,7) = 1.0
  190    CONTINUE
  
+c     compute derivatives of square root of conveyance, alpha, beta, 
+c     da, and dq
+
+        call xsecfit
+     I            (STDOUT, 0,
+     M             NDEPU, XSTU, 
+     o             khflag, alphaflag, betaflag, maflag, mqflag)
+
         CALL TABOUT
      I             (STDOUT, STDTAB, TABU, STATU, ZMINU, 0, SAVOPT,
-     I              OUTOPT, BETOPT,
+     I              OUTOPT, BETOPT, zone, hgrid, vdatum, unitsys, basis,
+     i                    khflag, alphaflag, betaflag, maflag, mqflag,
      M              NDEPU, XSTU, FTP)
  200  CONTINUE
       RETURN
@@ -1084,6 +1184,7 @@
       INCLUDE 'xscomu.cmn'
       INCLUDE 'nrdzcm.cmn'
       INCLUDE 'epscom.cmn'
+      INCLUDE 'xtadd.cmn'
  
 C     + + + LOCAL VARIABLES + + +
       INTEGER I, ISTART, IT, ITEND, J, JE, JS, NPIPES, NPNT, NSIDES,
@@ -1092,14 +1193,19 @@
      A     NMUD(PMXSUB), RIGHT, ROUGH(PMXSUB), WSLOT, XOFF, ZEPS,
      B     ZMUD(PMXSUB), ZMUD2, ZOFF
       CHARACTER BETOPT*8, CIN*64, CZEPS*10, LINE*80, MONTON*8, OUTOPT*8,
-     A          SAVOPT*8, TABID*16
+     A          SAVOPT*8,
+     a khflag(pmxpnt)*1, alphaflag(pmxpnt)*1, betaflag(pmxpnt)*1,
+     b     maflag(pmxpnt)*1, mqflag(pmxpnt)*1,
+     c     zone*8, hgrid*8, vdatum*8, unitsys*8, basis*8
+
+
  
 C     + + + INTRINSICS + + +
       INTRINSIC ABS, MAX, MIN
  
 C     + + + EXTERNAL NAMES + + +
       INTEGER LENSTR
-      EXTERNAL CXSTAB, INL80, INL80B, MKPIPE, SETMUD, SETOPT, TABCHK,
+      EXTERNAL CXSTAB, inline, inlineb, MKPIPE, SETMUD, SETOPT, TABCHK,
      A         TABOUT, LENSTR, READ_TABID_PLUS
  
 C     + + + INPUT FORMATS + + +
@@ -1132,10 +1238,12 @@
  92   FORMAT('0*WRN:549* BETA OPTION MUST BE "OLDBETA".  THE OPTION',
      A   ' HAS BEEN RESET TO "OLDBETA".')
 C***********************************************************************
+C     26 October 2006, ddf: Set the gisid to blank.  Uninitialized otherwise!
+      gisid = ' '
 C     ALLOW AUTOMATIC EXTENSION
       EXTEND = 1
  
-      CALL INL80
+      CALL inline
      I          (STDIN, STDOUT,
      O           LINE)
       CALL READ_TABID_PLUS
@@ -1164,13 +1272,23 @@
  
       TABU = TAB
  
-      CALL INL80
+c     Get location items that may be present. If they are not present
+c     they will be set to default values.  The default requests FEQUTL
+c     to omit the items. 
+      call GET_lctn_ITEMS(STDIN, STDOUT,
+     M                   EFLAG)
+      call  SET_lctn_ITEMS(
+     O             zone, hgrid, vdatum, unitsys, basis, easting,
+     O             northing)
+
+
+      CALL inline
      I          (STDIN, STDOUT,
      O           LINE)
       READ(LINE,4,ERR=991) NSIDES
       WRITE(STDOUT,54) NSIDES
  
-      CALL INL80
+      CALL inline
      I          (STDIN, STDOUT,
      O           LINE)
       READ(LINE,6,ERR=991) WSLOT
@@ -1178,13 +1296,13 @@
  
       SLOT = WSLOT
  
-      CALL INL80
+      CALL inline
      I          (STDIN, STDOUT,
      O           LINE)
       READ(LINE,6,ERR=991) HSLOT
       WRITE(STDOUT,58) HSLOT
  
-      CALL INL80
+      CALL inline
      I          (STDIN, STDOUT,
      O           LINE)
       READ(LINE,7,ERR=991) NPIPES, CZEPS
@@ -1209,19 +1327,19 @@
         JE = JS + 5
         JE = MIN(JE, NPIPES)
  
-        CALL INL80
+        CALL inline
      I            (STDIN, STDOUT,
      O             LINE)
         READ(LINE,8,ERR=991) (DIAM(J),J=JS,JE)
         WRITE(STDOUT,70) (DIAM(J), J=JS,JE)
  
-        CALL INL80
+        CALL inline
      I            (STDIN, STDOUT,
      O             LINE)
         READ(LINE,9,ERR=991) (BOTTOM(J), J=JS,JE)
         WRITE(STDOUT,72) (BOTTOM(J), J=JS,JE)
  
-        CALL INL80
+        CALL inline
      I            (STDIN, STDOUT,
      O             LINE)
         READ(LINE,10,ERR=991) (ROUGH(J), J=JS,JE)
@@ -1230,7 +1348,7 @@
 C       Look ahead for the mudline input line.  If not there, backspace
 C       the input file and set the mud line level to null.
  
-        CALL INL80B
+        CALL inlineb
      I            (STDIN, STDOUT,
      O             LINE)
         IF(LINE(1:4).NE.'MUDL') THEN
@@ -1243,7 +1361,7 @@
 C         READ THE MUDLINE LEVEL
           READ(LINE,12,ERR=991) (ZMUD(J), J=JS,JE)
           WRITE(STDOUT,76) (ZMUD(J), J=JS,JE)
-          CALL INL80
+          CALL inline
      I              (STDIN, STDOUT,
      O               LINE)
           READ(LINE,14,ERR=991) (NMUD(J), J=JS,JE)
@@ -1298,7 +1416,7 @@
         DO 400 J=1,NPNT
           SBU(ISTART+J-1) = I
           LSNU(ISTART+J-1) = ROUGH(I)
-          SNU(I) = 1.
+          SNU(ISTART+J-1) = 1.
  400    CONTINUE
  
         IF(ZMUD(I).GT.0.0) THEN
@@ -1323,15 +1441,15 @@
  
  500  CONTINUE
  
-      WRITE(STDOUT,*) ' '
-C      WRITE(STDOUT,*) ' DUMP OF CROSS-SECTION BOUNDARY'
-C      WRITE(STDOUT,*) ' '
-C
-C      DO 1000 I=1,NPNTU
-C        WRITE(STDOUT,90) XU(I), ZU(I), SBU(I), LSNU(I)
-C90    FORMAT(F10.4,F10.4,I5, F10.4)
-C1000  CONTINUE
-C      WRITE(STDOUT,*) ' '
+c      WRITE(STDOUT,*) ' '
+c      WRITE(STDOUT,*) ' DUMP OF CROSS-SECTION BOUNDARY'
+c      WRITE(STDOUT,*) ' '
+c
+c      DO 1000 I=1,NPNTU
+c        WRITE(STDOUT,90) XU(I), ZU(I), SBU(I), LSNU(I), SNU(I)
+c90    FORMAT(F10.4,F10.4,I5, F10.4,f10.2)
+c1000  CONTINUE
+c      WRITE(STDOUT,*) ' '
  
  
       NAVMU = 0
@@ -1404,9 +1522,18 @@
           XSTU(J,7) = 1.0
  190    CONTINUE
  
+c     compute derivatives of square root of conveyance, alpha, beta, 
+c     da, and dq
+
+        call xsecfit
+     I            (STDOUT, 0,
+     M             NDEPU, XSTU, 
+     o             khflag, alphaflag, betaflag, maflag, mqflag)
+
         CALL TABOUT
      I             (STDOUT, STDTAB, TABU, STATU, ZMINU, 0, SAVOPT,
-     I              OUTOPT, BETOPT,
+     I              OUTOPT, BETOPT, zone, hgrid, vdatum, unitsys, basis,
+     i                    khflag, alphaflag, betaflag, maflag, mqflag,
      M              NDEPU, XSTU, FTP)
  200  CONTINUE
       RETURN
@@ -1443,20 +1570,24 @@
       INCLUDE 'arsize.prm'
       INCLUDE 'xscomu.cmn'
       INCLUDE 'nrdzcm.cmn'
- 
+      INCLUDE 'xtadd.cmn'
 C     + + + LOCAL VARIABLES + + +
       INTEGER I, ISTART, J, NSIDES, TAB
       REAL D, HSLOT, LEFT, MAXK, N, NMUD, RIGHT, WSLOT, XOFF, ZEPS,
      A     ZMUD, ZOFF
       CHARACTER BETOPT*8, CIN*64, LINE*80, MONTON*8, OUTOPT*8, SAVOPT*8,
-     A          TABID*16
+     a khflag(pmxpnt)*1, alphaflag(pmxpnt)*1, betaflag(pmxpnt)*1,
+     b     maflag(pmxpnt)*1, mqflag(pmxpnt)*1,
+     c     zone*8, hgrid*8, vdatum*8, unitsys*8, basis*8
+     
+
  
 C     + + + INTRINSICS + + +
       INTRINSIC ABS
  
 C     + + + EXTERNAL NAMES + + +
       INTEGER LENSTR
-      EXTERNAL CXSTAB, INL80, INL80B, MKPIPE, SETMUD, SETOPT, TABCHK, 
+      EXTERNAL CXSTAB, inline, inlineb, MKPIPE, SETMUD, SETOPT, TABCHK, 
      A         TABOUT, READ_TABID_PLUS, LENSTR
  
 C     + + + INPUT FORMATS + + +
@@ -1475,20 +1606,22 @@
  56   FORMAT(' WIDTH OF SLOT=',F7.4)
  58   FORMAT(' HEIGHT OF SLOT=',F7.0)
  60   FORMAT(' MANNING N=',F7.3)
- 62   FORMAT('0*WRN:511* NUMBER OF SIDES=',I3,' TOO SMALL.',
+ 62   FORMAT(/,'*WRN:511* NUMBER OF SIDES=',I3,' TOO SMALL.',
      A       ' NSIDES RESET TO 10')
- 64   FORMAT('0*WRN:512* NUMBER OF SIDES=',I4,' TOO LARGE.',
+ 64   FORMAT(/,'*WRN:512* NUMBER OF SIDES=',I4,' TOO LARGE.',
      A       ' NSIDES RESET TO',I5)
- 66   FORMAT('0*WRN:549* BETA OPTION MUST BE "OLDBETA".  THE OPTION',
+ 66   FORMAT(/,'*WRN:549* BETA OPTION MUST BE "OLDBETA".  THE OPTION',
      A   ' HAS BEEN RESET TO "OLDBETA".')
- 68   FORMAT('0MUD LINE LEVEL=',F10.3)
- 70   FORMAT('0MUD LINE N=',F10.3)
+ 68   FORMAT(' MUD LINE LEVEL=',F10.3)
+ 70   FORMAT(' MUD LINE N=',F10.3)
 C***********************************************************************
+C     26 October 2006, ddf: Set the gisid to blank.  Uninitialized otherwise!
+      gisid = ' '
 C     ALLOW AUTOMATIC EXTENSION
  
       EXTEND = 1
  
-      CALL INL80
+      CALL inline
      I          (STDIN, STDOUT,
      O           LINE)
       CALL READ_TABID_PLUS
@@ -1517,19 +1650,28 @@
      M                         TAB, TABDIR, EFLAG)
       TABU = TAB
  
-      CALL INL80
+c     Get location items that may be present. If they are not present
+c     they will be set to default values.  The default requests FEQUTL
+c     to omit the items. 
+      call GET_lctn_ITEMS(STDIN, STDOUT,
+     M                   EFLAG)
+      call  SET_lctn_ITEMS(
+     O             zone, hgrid, vdatum, unitsys, basis, easting,
+     O             northing)
+
+      CALL inline
      I          (STDIN, STDOUT,
      O           LINE)
       READ(LINE,2,ERR=991) D
       WRITE(STDOUT,52) D
  
-      CALL INL80
+      CALL inline
      I          (STDIN, STDOUT,
      O           LINE)
       READ(LINE,4,ERR=991) NSIDES
       WRITE(STDOUT,54) NSIDES
  
-      CALL INL80
+      CALL inline
      I          (STDIN, STDOUT,
      O           LINE)
       READ(LINE,6,ERR=991) WSLOT
@@ -1537,13 +1679,13 @@
  
       SLOT = WSLOT
  
-      CALL INL80
+      CALL inline
      I          (STDIN, STDOUT,
      O           LINE)
       READ(LINE,6,ERR=991) HSLOT
       WRITE(STDOUT,58) HSLOT
  
-      CALL INL80
+      CALL inline
      I          (STDIN, STDOUT,
      O           LINE)
       READ(LINE,8,ERR=991) N
@@ -1553,7 +1695,7 @@
 C     LOOK AHEAD FOR THE MUDLINE INPUT LINE.  IF NOT THERE, BACKSPACE THE
 C     INPUT FILE AND SET THE MUD LINE LEVEL TO NULL.
  
-      CALL INL80B
+      CALL inlineb
      I          (STDIN, STDOUT,
      O           LINE)
       IF(LINE(1:4).NE.'MUDL') THEN
@@ -1564,7 +1706,7 @@
 C       READ THE MUDLINE LEVEL
         READ(LINE,10,ERR=991) ZMUD
         WRITE(STDOUT,68) ZMUD
-        CALL INL80
+        CALL inline
      I            (STDIN, STDOUT,
      O             LINE)
         READ(LINE,12,ERR=991) NMUD
@@ -1664,9 +1806,19 @@
  
       IF(EFLAG.NE.0.OR.TABU.EQ.0) GOTO 200
 C        IF(ABS(ZMINU).LT.0.001) ZMINU = 0.0
+
+c     compute derivatives of square root of conveyance, alpha, beta, 
+c     da, and dq
+
+          call xsecfit
+     I            (STDOUT, 0,
+     M             NDEPU, XSTU, 
+     o             khflag, alphaflag, betaflag, maflag, mqflag)
+
         CALL TABOUT
      I             (STDOUT, STDTAB, TABU, STATU, ZMINU, 0, SAVOPT,
-     I              OUTOPT, BETOPT,
+     I              OUTOPT, BETOPT, zone, hgrid, vdatum, unitsys, basis,
+     i                    khflag, alphaflag, betaflag, maflag, mqflag,
      M              NDEPU, XSTU, FTP)
  200  CONTINUE
       RETURN
@@ -2000,12 +2152,12 @@
 C     Z      - Elevation at points on cross section boundary
 
 C     + + + LOCAL VARIABLES + + +
-      INTEGER I, N
-
-      REAL H, DIV
+      INTEGER I, II, N
+
+      REAL H, DIV, PI
 
 C     + + + DATA INITIALIZATIONS + + +
-      DATA N/18/
+      DATA N/18/, PI/3.14159265/
 
 C***********************************************************************
 C     ADD AN EXTRA POINT WHICH IS LATER DELETED TO REMOVE NOISE FROM
@@ -2027,8 +2179,13 @@
       H = RISE - WSLOT
       DIV = FLOAT(N-1) 
       DO 100 I=1,N
+        ii = n - i + 1
         X(I+3) = -0.5*SPAN 
         Z(I+3) = H*(1.0 - SQRT(FLOAT(I-1)/DIV)) + ZOFF
+C                    3        2       1    1      1   12    2       1   123
+C        Z(I+3) = 0.5*(h -h*cos(pi*real(2*ii-1)/real(2*n))/
+C     a       cos(pi/real(2*n))) + ZOFF
+C        WRITE(*,'(I5,F10.3)') I, Z(I+3)
 100   CONTINUE
 
 C     Flip to the right-hand side to complete the perimeter
